File: 2024/04/day_04.jl
using Printf

mutable struct Solution
    function Solution()
        new()
    end
end


# function matrix_search(solution::Solution, matrix::[Vector[Char]], search_dirs::Vector{Vector{Tuple(Int, Int)}}, word::String)::Int
function matrix_search(solution::Solution, matrix::Vector{Vector{Char}}, search_dirs::Vector{Vector{Tuple{Int, Int}}}, word::String)::Int
    matrix_x = length(matrix[1])
    matrix_y = length(matrix)
    count = 0
    for i in eachindex(search_dirs)
        success = true
        for j in eachindex(word)
            if search_dirs[i][j][1] < 1 || search_dirs[i][j][1] > matrix_x || search_dirs[i][j][2] < 1 || search_dirs[i][j][2] > matrix_y
                success = false
                break
            end
            if matrix[search_dirs[i][j][2]][search_dirs[i][j][1]] != word[j]
                success = false
                break
            end
        end
        if success
            count = count + 1
        end
    end
    return count
end

function find_words(solution::Solution, lines::Vector{String}, word::String)::Int
    total = 0
    matrix = [Vector{Char}(undef, length(lines[1])) for _ in 1:length(lines)]
    for (i, line) in enumerate(lines)
        for (j, char) in enumerate(line)
            matrix[i][j] = char
        end
    end

    # Perform a matrix search for the word
    for i in 1:length(matrix)
        for j in 1:length(matrix[1])
            if matrix[i][j] == word[1]
                # create a set of arrays of coordinates to search
                search_dirs = [Vector{Tuple{Int, Int}}(undef, length(word)-1) for _ in 1:8]
                for k in 1:length(word)-1
                    # we store (x,y) coordinates
                    search_dirs[1][k] = (j+k, i) # right
                    search_dirs[2][k] = (j-k, i) # left
                    search_dirs[3][k] = (j, i+k) # down
                    search_dirs[4][k] = (j, i-k) # up
                    search_dirs[5][k] = (j+k, i+k) # down-right
                    search_dirs[6][k] = (j-k, i-k) # up-left
                    search_dirs[7][k] = (j-k, i+k) # down-left
                    search_dirs[8][k] = (j+k, i-k) # up-right
                end

                total = total + matrix_search(solution, matrix, search_dirs, word[2:end])
            end
        end
    end
    return total
end

function find_words_x(solution::Solution, lines::Vector{String}, word::String)::Int
    if length(word)%2 != 1
        throw(ArgumentError("The length of the word must be odd"))
    end

    total = 0
    matrix = [Vector{Char}(undef, length(lines[1])) for _ in 1:length(lines)]
    for (i, line) in enumerate(lines)
        for (j, char) in enumerate(line)
            matrix[i][j] = char
        end
    end


    center_letter = word[Int((length(word)+1)/2)]
    start_offset = Int((length(word)-1)/2)
    # Perform a matrix search for the word
    for i in 1:length(matrix)
        for j in 1:length(matrix[1])
            if matrix[i][j] == center_letter
                # create a set of arrays of coordinates to search
                search_dirs1 = [Vector{Tuple{Int, Int}}(undef, length(word)) for _ in 1:2]
                search_dirs2 = [Vector{Tuple{Int, Int}}(undef, length(word)) for _ in 1:2]
                for k in 1:length(word)
                    # we store (x,y) coordinates in two sets - this allows for word to be a palindrome
                    search_dirs1[1][k] = (j+(k-1-start_offset), i+(k-1-start_offset)) # down-right
                    search_dirs1[2][k] = (j-(k-1-start_offset), i-(k-1-start_offset)) # up-left

                    search_dirs2[1][k] = (j+(k-1-start_offset), i-(k-1-start_offset)) # up-right
                    search_dirs2[2][k] = (j-(k-1-start_offset), i+(k-1-start_offset)) # down-left
                end

                if matrix_search(solution, matrix, search_dirs1, word) > 0 &&
                   matrix_search(solution, matrix, search_dirs2, word) > 0
                    total = total + 1
                end
            end
        end
    end
    return total
end

function read_data(fn::String)::Vector{String}
    return readlines(fn)
end

if abspath(PROGRAM_FILE) == @__FILE__
    data = read_data("input.txt")

    s = Solution()
    basic_words = find_words(s, data, "XMAS")
    x_words = find_words_x(s, data, "MAS")
    
    println(basic_words)
    println(x_words)
end

File: 2024/04/day_04.py
from typing import List

class Solution:
    def __init__(self):
        pass

    def find_words_x(self, words: List[str], search_word: str) -> int:
        assert len(search_word)%2 == 1, "search word must be odd length"
        search_letter = search_word[len(search_word)//2]
        count = 0
        for i in range(len(words)):
            indices = [index for index, letter in enumerate(words[i]) if letter == search_letter]
            for idx in indices:
                valid = True
                word1 = search_letter
                word2 = search_letter
                for offset in range(1, len(search_word)//2+1):
                    if i - offset >= 0 and i + offset < len(words) and idx - offset >= 0 and idx + offset < len(words[i]):
                        word1 = words[i-offset][idx-offset] + word1 + words[i+offset][idx+offset]
                        word2 = words[i-offset][idx+offset] + word2 + words[i+offset][idx-offset]
                    else:
                        valid = False
                        break
                if valid and (search_word in word1 or search_word in word1[::-1]) \
                         and (search_word in word2 or search_word in word2[::-1]):
                    count = count + 1
        return count
                
    def find_words(self, words: List[str], search_word: str) -> int:
        # simple method - use python string processing
        count = 0
        # print("")

        # serach each row
        for line in words:
            # print(line + " " + line[::-1])
            count = count + line.count(search_word)
            count = count + line[::-1].count(search_word)
        
        # search each column
        for i in range(0, len(words[0])):
            column = "".join([x[i] for x in words])
            count = count + column.count(search_word)
            count = count + column[::-1].count(search_word)
        
        # print("")
        # search down right diagonal
        for row in range(0, len(words)):
            search_elems = [0]
            if row == 0:
                search_elems = range(0, len(words[row]))
            for col in search_elems:
                i, j = row, col
                word = ""
                bottom_word = ""
                while True:
                    word = word + words[i][j]
                    bottom_word = bottom_word + words[-1*(i+1)][j]
                    # print(-1*(i+1), j, bottom_word, words[-1*(i+1)][j])
                    i = i + 1
                    j = j + 1
                    if i >= len(words) or j >= len(words[i]):
                        break
                # print(row, col)
                # inc_count =  word.count(search_word) + word[::-1].count(search_word)
                # print(word + " " + word[::-1] + " " + str(inc_count))

                # inc_count =  bottom_word.count(search_word) + bottom_word[::-1].count(search_word)
                # print(bottom_word + " " + bottom_word[::-1] + " " + str(inc_count))

                count = count + word.count(search_word)
                count = count + word[::-1].count(search_word)

                count = count + bottom_word.count(search_word)
                count = count + bottom_word[::-1].count(search_word)
        
        # print("\n\n")
        return count
    

def main(fn: str):
    with open(fn) as f:
        data = [x.strip() for x in f.readlines()]
        s = Solution()
        result = s.find_words(data, "XMAS")
        print(result)
        result = s.find_words_x(data, "MAS")
        print(result)

if __name__ == "__main__":
    main("input.txt")